renderdoc/driver/ihv/intel/intel_counters.cpp

```diff
@@ -51,31 +50,35 @@ IntelCounters::~IntelCounters()
     FreeLibrary(m_MDLibraryHandle);
 }
 
-bool IntelCounters::Init(void *pContext)
+void IntelCounters::Load()
 {
-  if(!pContext)
-    return false;
-
-  m_device = (ID3D11Device *)pContext;
-  m_device->GetImmediateContext(&m_deviceContext);
-  if(!m_deviceContext)
-    return false;
-
   m_MDLibraryHandle = LoadDynamicLibrary(L"igdmd64.dll", NULL, 0);
   if(!m_MDLibraryHandle)
-    return false;
+    return;
 
   OpenMetricsDevice = (OpenMetricsDevice_fn)GetProcAddress(m_MDLibraryHandle, "OpenMetricsDevice");
   if(!OpenMetricsDevice)
-    return false;
+    return;
 
   CloseMetricsDevice =
       (CloseMetricsDevice_fn)GetProcAddress(m_MDLibraryHandle, "CloseMetricsDevice");
   if(!CloseMetricsDevice)
+    return;
+
+  OpenMetricsDevice(&m_metricsDevice);
```

Comment:
`IntelCounters::Load` is called unconditionally when starting the D3D11 replay, but it only gets closed again if we actually create an `IntelCounters` and it either fails to initialise or we keep it around and shut it down with the replay.

That means if `IntelCounters::Load` is called but we don't detect an Intel GPU, the matching `CloseMetricsDevice` will never be called, but we'll still call `OpenMetricsDevice` again next time we come into `IntelCounters::Load`. You said it would be safe to have one metrics device for many D3D11 devices, so I think it would be better to have the metrics device be initialised lazily - i.e here we only do this if `m_metricsDevice` is non-NULL - then it can be closed once at program shutdown.